Profile epic

roma685235423 opened the pull request:
Архитектура MVP.

Декомпозиция:
https://github.com/miamib34ch/YP-CP-FakeNFT/blob/profile_epic/FakeNFT/Decomposition/Profile.md

Таск трекер:
https://github.com/users/miamib34ch/projects/1/views/1

Скринкаст:
https://drive.google.com/file/d/19xHCsXBKaqnq3i1rIFSEG5HX4isIPZ6F/view?usp=share_link

roma685235423 committed:
[Профиль] Сверстать экран "Профиль"

Создал ProfileViewController, ячейку ProfileTableViewCell для таблицы навигации информации профиля.

roma685235423 committed:
[Профиль] Сверстать экран "Профиль"

Создал ProfileNetworkClient и ProfileViewPresenter. Реализовал с их помощью загрузку данных профиля и фото. В ProfileViewController добавил activityIndicator и логику его отображения во время загрузки фото и скрытия после в ProfileViewPresenter.

roma685235423 committed:
[Профиль] Сверстать экран "Профиль"

Реализовал загрузку и отображение всех данных пользователя и изменение заголовков ячеек в зависимости от ответа

roma685235423 committed:
[Профиль] Сверстать экран "Редактирование профиля"

Сверстал экран редактирования профиля, создал кастомные поля редактирования данных, добавил алерт с полем ввода ссылки и алерт неуспешной проверки ссылки, так же произвёл небольшие изменения в именах сущностей.

roma685235423 committed:
[Профиль] Сверстать экран "Редактирование профиля"

Реализовал передачу редактированных данных профиля в экран профиля для их отправки с помощью сетевого слоя презентера

roma685235423 committed:
[Профиль] Обновление данных на сервере.

Реализована логика отправки обновлённых данных на сервер после редактирования на экране. С помощью ProcessHUD добавлен индикатор загрузки, и блокировка интерфейса.

roma685235423 committed:
Merge pull request #58 from roma685235423/profile_epic

Profile epic

roma685235423 committed:
Merge branch 'base' into profile_epic

roma685235423 committed:
[Профиль] Создать веб-вью для просмотра сайта разработчика

Добавил UserAboutWebView и реализовал модальное открытие веб вью при нажатии на лэйбл сайта или ячейку "О разработчике"

roma685235423 committed:
[Профиль] Сверстать экран "Мои NFT"

Приступил к вёрстке экрана "Мои NFT"

roma685235423 committed:
[Профиль] Сверстать ячейку для UITableView экрана "Мои NFT"

Приступил к созданию ячейки. Создал модели ответа для сущности автор, NFT, а так же для ячейки. Так же в MyNFTsViewController добавил TableView.

roma685235423 committed:
[Профиль] Реализовать логику загрузки данных UI элементов экрана "Мои NFT".

Преступил к созданию сервиса загрузки данных для экрана и презентера.

roma685235423 committed:
[Профиль] Сверстать ячейку для UICollectionView экрана "Избранные NFT"

Сверстал ячейку, UICollectionView. В процессе ViewController и Presenter

roma685235423 committed:
[Профиль] Реализовать загрузку данных "Мои NFT".

Реализовал загрузку данных и их обработку для слоя представления экрана "Мои NFT".

roma685235423 committed:
[Профиль] Сверстать ячейку для UITableView экрана "Мои NFT"

Продолжил вёрстку ячейки и экрана "Мои NFT", для этого внёс изменения в презентер

roma685235423 committed:
[Профиль] Сверстать ячейку для UITableView экрана "Мои NFT"

Окончил вёрстку ячейки и экрана. Добавил плэйсхолдер и алерт презентер

roma685235423 committed:
[Профиль] Сверстать экран "Редактирование профиля"

Исправил ошибку, препятствующую отправке изменённой сслыкли фото профиля

roma685235423 committed:
[Профиль] Сверстать ячейку для UITableView экрана "Мои NFT"

Добавил расширение к UIStackViewController с методом, наполняющим стак изображениями звёзд в зависимости от рейтинга.

So1ovyov commented on FakeNFT/Epics/Profile/Models/AuthorResponseModel.swift:
```diff
@@ -0,0 +1,11 @@
+import Foundation
+
+struct AuthorResponseModel: Codable {
```
Было бы круто добавить документацию к структуре и её свойствам.

roma685235423 replied:
Готово.

So1ovyov commented on FakeNFT/Epics/Profile/Models/EditableProfileModel.swift:
```diff
@@ -0,0 +1,12 @@
+import UIKit
+
+struct EditableProfileModel {
```
Было бы круто добавить документацию к структуре и её свойствам.

roma685235423 replied:
Добавил

roma685235423 replied:
Готово.

So1ovyov commented on FakeNFT/Epics/Profile/Models/MyNFTModel.swift:
```diff
@@ -0,0 +1,10 @@
+import Foundation
+
+struct MyNFCModel: Codable {
```
Было бы круто добавить документацию к структуре и её свойствам.

roma685235423 replied:
Добавил

roma685235423 replied:
Готово.

So1ovyov commented on FakeNFT/Epics/Profile/Models/MyNFTPresentationModel.swift:
```diff
@@ -0,0 +1,10 @@
+import Foundation
+
+struct MyNFTPresentationModel {
```
Было бы круто добавить документацию к структуре и её свойствам.

roma685235423 replied:
Готово.

So1ovyov commented on FakeNFT/Epics/Profile/Models/NFTResponseModel.swift:
```diff
@@ -0,0 +1,12 @@
+import Foundation
+
+struct NFTResponseModel: Decodable {
```
Было бы круто добавить документацию к структуре и её свойствам.

roma685235423 replied:
Добавил

roma685235423 replied:
Готово.

So1ovyov commented on FakeNFT/Epics/Profile/Models/ProfileResponseModel.swift:
```diff
@@ -0,0 +1,11 @@
+import Foundation
+
+struct ProfileResponseModel: Codable {
```
Было бы круто добавить документацию к структуре и её свойствам.

roma685235423 replied:
Добавил

roma685235423 replied:
Готово.

So1ovyov commented on FakeNFT/Epics/Profile/Presentation/FeaturedNFTs/CollectionParameters.swift:
```diff
@@ -0,0 +1,19 @@
+import UIKit
+
+struct FeaturedCollectionParameters {
```
Было бы круто добавить документацию к структуре и её свойствам.

roma685235423 replied:
Готово.

So1ovyov commented on FakeNFT/Epics/Profile/Presentation/FeaturedNFTs/CollectionParameters.swift:
```diff
@@ -0,0 +1,19 @@
+    let paddingWidth: CGFloat
+    
+    // MARK: - Life cycle
+    init(cellCount: Int, leftInset: CGFloat, rightInset: CGFloat, cellSpacing: CGFloat) {
```
Между MARK-ой и сущностью хорошо ставить пустую строку для разделения.
Большие конструкции можно разделять следующим образом:
```
    // MARK: - Life cycle
    
    init(
        cellCount: Int,
        leftInset: CGFloat,
        rightInset: CGFloat,
        cellSpacing: CGFloat
    ) {
        self.cellCount = cellCount
        self.leftInset = leftInset
        self.rightInset = rightInset
        self.cellSpacing = cellSpacing
        self.paddingWidth = leftInset + rightInset + CGFloat(cellCount - 1) * cellSpacing
    }

```

roma685235423 replied:
Поправил

So1ovyov commented on FakeNFT/Epics/Profile/Presentation/FeaturedNFTs/FeaturedNFTCell.swift:
```diff
@@ -0,0 +1,116 @@
+    // MARK: - Private methods
+    private func addingUIElements() {
+        [nftImage, aboutNFCStack, likeButton].forEach{
+            $0.translatesAutoresizingMaskIntoConstraints = false
```
Можно перенести эти строки в сами элементы выше, где ты их настраиваешь.

roma685235423 replied:
Решил не выходить за рамки общей концепции

So1ovyov commented on FakeNFT/Epics/Profile/Presentation/FeaturedNFTs/FeaturedNFTsCell.swift:
```diff
@@ -0,0 +1,115 @@
+        label.text = "0 ETH"
+        return label
+    }()
+
```
Лишняя пустая строка.

roma685235423 replied:
Удалил

So1ovyov commented on FakeNFT/Epics/Profile/Presentation/FeaturedNFTs/FeaturedNFTsViewController.swift:
```diff
@@ -0,0 +1,135 @@
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        view.backgroundColor = .ypWhite
+//        presenter?.viewDidLoad()
```
Закомментированный код можно удалить.

roma685235423 replied:
Раскомментировал) Эта строка нужна

So1ovyov commented on FakeNFT/Epics/Profile/Presentation/MyNFTs/MyNFTsPresenter.swift:
```diff
@@ -0,0 +1,161 @@
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
```
Лишние пустые строки можно удалить.

roma685235423 replied:
Удалил

So1ovyov commented on FakeNFT/Epics/Profile/Presentation/MyNFTs/MyNFTsPresenter.swift:
```diff
@@ -0,0 +1,161 @@
+    func viewDidLoad() {
+        getNFTResponceModels(for: profile.nfts)
+    }
+    
```
Лишние пустые строки можно удалить.

roma685235423 replied:
Удалил

So1ovyov commented on FakeNFT/Epics/Profile/Presentation/ProfileEdit/ProfileEditViewController.swift:
```diff
@@ -0,0 +1,281 @@
+    private let descriptionLabel = EditedParameterLabel.init(text: NSLocalizedString("profile.description", comment: ""))
+    private let siteLabel = EditedParameterLabel.init(text: NSLocalizedString("profile.site", comment: ""))
+    
+    // swiftlint:disable:next trailing_closure
```
А точно ли оно ругается ? У меня в проекте нет ворнинга. Нет ли возможности без отключения линтера это исправить?

So1ovyov commented on FakeNFT/Epics/Profile/Protocols/ProfileProtocols.swift:
```diff
@@ -0,0 +1,40 @@
+import UIKit
+
+// MARK: - ProfileViewControllerProtocol
+protocol ProfileViewControllerProtocol: AnyObject {
```
Можно вынести в отдельные файлы эти протоколы.

roma685235423 replied:
Вынес

roma685235423 committed:
[Профиль] Сверстать UICollectionView экрана "Избранные NFT"

Реализовал загрузку данных, вызов алерта с возможностью повторного запроса, привёл ячейку к необходимому виду, добавил плейсхолдер и логику удаления ячейки из коллекции:

roma685235423 committed:
[Профиль] Логика загрузки данных "Избранные NFT".

Реализовал логику обновления профиля на сервере с учётом изменения массива likes профиля пользователя и обновление экрана профиля после закрытия экрана "Избранные NFT"

roma685235423 committed:
[Профиль] Сортировка содержимого экрана "Мои NFT"

Реализовал сортировку содержимого UITableView, а так же вызов алерта для выбора опции сортировки

roma685235423 committed:
[Профиль] Поправил форматирование в файлах эпика.

roma685235423 committed:
[Профиль] Изменил имена файлов для решения конфликтов

roma685235423 committed:
Merge branch 'base' into profile_epic

roma685235423 committed:
[Профиль] Добавил алерт презентер на экран профиля

miamib34ch commented:
Привет! Посмотрел твой эпик, заметил моменты, которые нужно исправить: 

- Текст описания автора не весь входит на экран, скорее всего, это из-за того, что высота фиксированная
- Имя тоже не всё входит, если длинное, может добавить, вторую строку?
- При редактировании профиля нельзя убрать клавиатуру, из-за чего изменить сайт автора нельзя без перезахода + мешает виденью контента
- Сортировка по-умолчанию неправильная, должна быть по рейтингу
- Выбранная сортировка не сохраняется на устройстве
- Цвет серой звездочки в темной теме неправильный (для бейз ветки пулл реквест висит с исправлением этого, надо принять и замерджить)
- Ячейки мои нфт и избранные нфт не соотвествуют макетам, нужно шрифты поменять
- Расстояние между ячейками неправильное в избранные нфт

Также мне в моём ревью от Практикума сообщили следующее: "вью должна сообщать презентеру о действиях юзера. вью не может говорить презентеру, что ему делать, потому что она пассивная. поэтому везде, где вью обращается с указаниями к презентеру, нужно заменить название метода на событие. Например: твоя вью загрузилась и вместо того, чтоб сказать презентеру, чтобы он загружал данные, она должна просто сказать: viewDidLoad. А презентер сам уже должен решить, что ему с этим событием делать", думаю в твоей работе можно это тоже поправить.  
Ещё хотел бы отметить, что функции, которые ничего не получают на вход, а только возвращают данные, можно переделать в автоматические свойства.

roma685235423 committed:
[Профиль] Добавил сохранение опции сортировки в UserDefaults

Добавил сущность UserDefaultsManager для работы с UserDefaults

roma685235423 committed:
[Профиль] Исправил замечания к внешнему виду ячеек.

Исправил шрифты лэйблов FavoriteNFTCell и MyNFTCell. Привёл к соответствию с макетом отступы между ячейками коллекции и их ширину.

roma685235423 committed:
[Профиль] Добавил изменение положения по оси Y

Для ProfileEditViewController при показе клавиатуры

roma685235423 committed:
Merge branch 'base' into profile_epic

roma685235423 committed:
Merge branch 'base' into profile_epic

roma685235423 committed:
Resolve merge conflicts in TabBar.swift and Image.swift

roma685235423 committed:
Merge branch 'profile_epic'

of https://github.com/miamib34ch/YP-CP-FakeNFT into profile_epic

roma685235423 committed:
[Профиль] Добавил документацию к моделям.

roma685235423 committed:
[Профиль] Исправление замечаний

Заменил функции, ничего не получающие на вход, а только возвращают данные, на автоматические свойства

roma685235423 committed:
[Профиль] Внес корректировки, в лэйблы профиля.

Заменил userDescriptionLabel: UILabel на userDescriptionTextView: UITextView для того, чтобы можно было скроллить текст, не прибегая к изменениям layout, в userNameLabel добавил вторую строку.

roma685235423 commented:
Богдан, большое спасибо за твои комментарии в ревью. Я уже внес все необходимые изменения согласно твоим рекомендациям. Выражаю отдельную благодарность за твои замечания и советы относительно функций, которые можно переделать в автоматические свойства. Это действительно ценная практика, способствующая улучшению структуры кода. Внесенные изменения в моей доработке полностью учтены, и я уверен, что это будет полезным опытом для будущих проектов.

roma685235423 committed:
Merge branch 'develop' into profile_epic

roma685235423 committed:
[Профиль] Добавил сервис для отправки лайков

roma685235423 committed:
Merge branch 'origin/profile_epic'  into profile_epic

roma685235423 committed:
[Профиль] Интегрировал общий сервис лайков.

roma685235423 committed:
Перенёс загрузку данных профиля в презентер экрана Профиль

roma685235423 committed:
[Профиль] Изменил название метода презентера профиля.

С viewDidLoad на viewWillAppear.

roma685235423 committed:
Вернул функцию в AppDelegate

func application(_: UIApplication, didFinishLaunchingWithOptions _: [UIApplication.LaunchOptionsKey: Any]?) -> Bool
